File: exact.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

func shortestTour(nodes [][]float64, presentNode int, visitedNodes int, lookupMatrix [][]float64) float64 {
	if visitedNodes == (1<<uint(len(nodes)) - 1) {
		return nodes[presentNode][0]
	}

	if lookupMatrix[presentNode][visitedNodes] != math.Inf(1) {
		return lookupMatrix[presentNode][visitedNodes]
	}
	//var tour []int
	for i := 0; i < len(nodes); i++ {
		if (i == presentNode) || ((visitedNodes & (1 << uint(i))) > 0) {
			continue
		}

		newDistance := nodes[presentNode][i] + shortestTour(nodes, i, visitedNodes|(1<<uint(i)), lookupMatrix)

		if newDistance < lookupMatrix[presentNode][visitedNodes] {
			//tour = append(tour, i)
			lookupMatrix[presentNode][visitedNodes] = newDistance
		}
	}
	//fmt.Println("shortestTour ", tour)
	return lookupMatrix[presentNode][visitedNodes]
}

func main() {
	var distances [][]float64
	file, err := os.Open("/home/raja/Downloads/programming/python/AntColonyOptimization-master/att48_d.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var row []float64
		line := strings.Split(scanner.Text(), "      ")
		for _, str := range line[1:] {
			dis, _ := strconv.ParseFloat(strings.Trim(str, " "), 10)
			if dis == 0.0 {
				row = append(row, math.Inf(1))
			} else {
				row = append(row, dis)

			}
		}
		distances = append(distances, row)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	//distances1 := [][]float64{{0, 20, 42, 25},
	//	{20, 0, 30, 34},
	//	{42, 30, 0, 10},
	//	{25, 34, 10, 0}}
	var distances1 [][]float64
	var lookupMatrix [][]float64
	for i := 0; i < 25; i++ {
		var temp []float64
		for j := 0; j < 25; j++ {
			temp = append(temp, distances[i][j])
		}
		distances1 = append(distances1, temp)
	}
	for i := 0; i < len(distances1); i++ {
		var temp []float64
		for j := 0; j < (1<<uint(len(distances1)))-1; j++ {
			temp = append(temp, math.Inf(1))
		}
		lookupMatrix = append(lookupMatrix, temp)
	}
	fmt.Println("nodes size:", len(distances1), " ", len(distances1[0]))
	fmt.Println("lookup size:", len(lookupMatrix), " ", len(lookupMatrix[0]))
	fmt.Println("shortest distance", shortestTour(distances1, 0, 1, lookupMatrix))
}

File: exact.rs
use std::f32::INFINITY;
use std::io::BufRead;
use std::io::BufReader;
use std::io::{BufWriter, Write};

fn shortest_tour(
    nodes: Vec<Vec<f64>>,
    present_node: i64,
    visited_nodes: i64,
    lookupMatrix: Vec<Vec<f64>>,
) {
    if visited_nodes == (1 << nodes.len()) - 1 {
        return nodes[present_node][0];
    }
    if lookupMatrix[present_node][visited_nodes] != INFINITY {
        return lookupMatrix[present_node][visited_nodes];
    }

    for i in 0..nodes.len() {
        if (i == present_node) || ((visited_nodes & (1 << i)) > 0) {
            continue;
        }
        let new_distance = nodes[present_node][i]
            + shortest_tour(nodes, i, visited_nodes | (1 << i), lookupMatrix);

        if new_distance < lookupMatrix[present_node][visited_nodes] {
            lookupMatrix[present_node][visited_nodes] = new_distance;
        }
    }
    return lookupMatrix[present_node][visited_nodes];
}

fn main() {
    let distances: Vec<Vec<f64>> = Vec::new();
    let mut file = BufReader::with_capacity(
        1000 * 1024,
        File::open(
            "/home/raja/Downloads/programming/python/AntColonyOptimization-master/att48_d.txt",
        )
        .unwrap(),
    );
    let mut line = String::new();
    while file.read_line(&mut line).unwrap() > 0 {
        let words: Vec<f64> = line.split("      ").parse().collect();
        distances.push(words);
    }
    let distances1: Vec<Vec<f64>> = Vec::new();
    let lookupMatrix: Vec<Vec<f64>> = Vec::new();
    for i in 0..24 {
        let temp: Vec<f64> = Vec::new();
        for j in 0..24 {
            temp.push(distances[i][j])
        }
        distances1.push(temp);
    }

    for i in 0..distances1.len() {
        let temp: Vec<f64> = Vec::new();
        for j in 0..((1 << distances1.len()) - 1) {
            temp.push(INFINITY)
        }
        lookupMatrix.push(temp);
    }

    println!("nodes size: {} {}", distances1.len(), distances.len());
    println!(
        "lookup size: {} {}",
        lookupMatrix.len(),
        lookupMatrix[0].len()
    );
    println!(
        "shortestTour {}",
        shortest_tour(distances1, 0, 1, lookupMatrix)
    );
}
